---
title: "Extended Data Fig 1"
author: "Regan"
date: "16/05/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script will generate the scRNA-seq data figures in extended data figure 1 from the dataset that has been pre-processed with scripts 1-6.


# Extended Data Fig. 1c - UMAPs of the dataset separated by sex

Load the sce
```{r}
rm(list = ls())

sce_all <- readRDS("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Cell Ranger 3.1.0/SCI/May 2020/BC/sce.RDS")
```

Generate UMAPs for each sex with greyed out background cells
```{r}
library(scater)
library(RColorBrewer)
library(Seurat)

# set the sex for subsetting the sce
sex <- c("F", "M") 

# choose how to colour and shape the UMAP
c.by <- "grey"
s.by <- NULL 
exp <- "logcounts"
#colours <- c(FF9200)  #c(brewer.pal(nlevels(sce_all$sex), "Accent"),"grey95") # c("#D53E4F", "#FF8D59","#E6F598","#1A9850","#3288BD", "#542788", "grey95")

# name for labelling the UMAP
name <- c.by

# load the sce and generate the UMAP for each timepoint
for (j in sex) {
  sce <- sce_all
  
  #set the colour category for the UMAP
  col.name <- "sex"
  col <- sce$sex 
  
  #set the colours
  if (j == "F") {
    colours <- c("#FF9200", "grey95", "grey95")
  } else {
    colours <- c( "grey95","#1CCB12", "grey95")
  }
  
  #set the colour category for the grey
  category <- j
  subset <- sce$sex
  grey <- subset != category
  
  # rename levels to include "grey" for those not of interest
  sce$grey <- factor(col, levels=c(levels(col), "grey"))
  sce$grey[grey] <- "grey"
  
  #sort the cells in order to have the grey cells last (and therefore underneath)
  sce <- sce[,rev(order(colnames(sce)[sce$grey])), ]
  
  # generate the UMAP for the timepoint
  p <- plotUMAP(sce, colour_by=c.by, shape_by=s.by, by_exprs_values=exp, point_size = 1) +  ggtitle(paste0(j))  + theme(plot.title=element_text(size=rel(2)))  + geom_point(aes(fill=sce$grey),colour= "grey95",  pch=21, stroke = .1, size =1.5) + scale_fill_manual(values=colours, drop=F) + theme(legend.position="none") # #+ xlim(-12.5,7.5) + ylim(-12.5, 12.5)
  
  # convert the UMAP dots to a single image (easier handling in Illustrator)
  AugmentPlot(plot=p, dpi=600)  
  
  # save the plot 
  ggsave(paste0("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Cell Ranger 3.1.0/SCI/May 2020/BC/Hamel et al 202X/S. Fig 1c/grey sex ",j,".pdf"), width=9, height=9)
}
```


# Extended Data Fig. 1d Comparison of percent of counts from immediate early genes (IEGs) from microglia across several scRNA-seq studies
- first we will load the datasets and apply the same QC thresholds to each
- then we will calculate the percent of counts from IEGs
- finally, we will plot the percent of IEGs counts in a violin plot

Initialise the function to prepare each single cell experiment using the same QC metrics
```{r}
rm(list = ls())
```

--------------------- SCI data ---------------------
Load the SCI data *before* low quality cluster removal and extract only the resident CNS data
```{r}
sce.SCI <- readRDS("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Cell Ranger 3.1.0/SCI/May 2020/BC QC clust NOT removed/sce all.RDS")

# extract only HC 
sce.SCI <- sce.SCI[,sce.SCI$stage == "HC"]
```

--------------------- Yang 2018 data ---------------------
Load the Yang 2018 data from the gene-counts matrix downloaded from the GEO: GSE101901
- this study investigated CNS cells post-traumatic brain injury using DropSeq
```{r}
library(SingleCellExperiment)

counts <- read.table("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Yang TBI snRNA data/GSE101901_DropSeqTBI.digital_expression.txt")

all.counts <- as.matrix(counts)
sce.Y <- SingleCellExperiment(list(counts=all.counts))
dim(sce.Y)
```

Isolate HC microglia, defined as C1q+ cells (by the authors of the Yang 2018 dataset)
- We confirmed the expression of microglia markers in this subpopulation (data not shown)
- microglia markers that we confirmed to be expressed in this subset: "Tmem119", "Iba1", "Itgam", "P2ry12", "P2ry13", "Gpr34", "Sparc", "Olfml3", "Hexb", "Fcrls", "Trem2", "C1qa","C1qb", "C1qc", "Sall1", "Ptprc", "Cx3cr1", "Adgre1", "Cd68", "Cd40"
```{r}
# isolate only HC cells
sce.Y <- sce.Y[,grep("Sham.*", colnames(sce.Y))]

# remove cells without C1qa 
sce.Y <- sce.Y[,counts(sce.Y["C1qa"]) > 0]
```

--------------------- Stevens 2019 data ---------------------
- this study enriched for microglia by FACS for Cd45-low Cd11b-high Cx3cr1-high cells before sequencing via 10X Chromium single cell 3' solution. We examined only their P100 dataset from healthy and LPC injured mice, as these samples were the most similar to our current study.
- note this dataset had mitochondrial genes removed
- note that their P30 dataset had similar IEG % but was strictly HC

Load the Stevens 2019 data from the sample-specific gene-counts matrices downloaded from the GEO: GSE121654
```{r}
library(stringr)
list <- c("GSM3442030_P100_male_3.dge.txt GSM3442026_P100_Male_1.dge.txt GSM3442026_P100_Male_1.dge.txt.gz GSM3442027_P100_Male_2.dge.txt GSM3442027_P100_Male_2.dge.txt.gz GSM3442028_P100_female_1.dge.txt GSM3442028_P100_female_1.dge.txt.gz GSM3442029_P100_female_2.dge.txt GSM3442029_P100_female_2.dge.txt.gz GSM3442030_P100_male_3.dge.txt.gz GSM3442031_P100_male_4.dge.txt GSM3442031_P100_male_4.dge.txt.gz GSM3442032_P100_female_3.dge.txt GSM3442032_P100_female_3.dge.txt.gz GSM3442033_P100_female_4.dge.txt GSM3442033_P100_female_4.dge.txt.gz GSM3442038_P100_M_A1.dge.txt GSM3442038_P100_M_A1.dge.txt.gz GSM3442039_P100_M_A2.dge.txt GSM3442039_P100_M_A2.dge.txt.gz GSM3442040_P100_M_B5.dge.txt GSM3442040_P100_M_B5.dge.txt.gz GSM3442041_P100_M_SALINE_A3.dge.txt GSM3442041_P100_M_SALINE_A3.dge.txt.gz GSM3442042_P100_M_SALINE_A5.dge.txt GSM3442042_P100_M_SALINE_A5.dge.txt.gz GSM3442043_P100_M_SALINE_B9.dge.txt GSM3442043_P100_M_SALINE_B9.dge.txt.gz")

# separates the strings by empty space 
list <- unlist(strsplit(list, split = " "))

#remove the zip files
list <- gsub(".*gz", "", list)

#remove the whitespace on each end
list <- trimws(list, which = "b", whitespace = "[ \t\r\n]")

#remove empties
list <- list[list!=""]

#load in and combine each sample
dirG <- "/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Stevens 2019/P100/"

# load the first sample
all.counts <- read.delim(paste0(dirG, list[1]))
rownames(all.counts) <- all.counts[,1]
all.counts <- all.counts[,-1]

list <- list[-1]
head(all.counts)

# load and combine the remaining samples
for (i in list) {
  #read in counts
  counts <- read.delim(paste0(dirG, i))
  rownames(counts) <- counts[,1]
  counts <- counts[,-1]
  
  # match the genes across samples by determining which sample contains more genes and taking the shorter of the two
  if (nrow(counts) > nrow(all.counts)) {
    counts <- counts[rownames(all.counts),]
    rownames <- rownames(counts)
  } else {
    all.counts <- all.counts[rownames(counts),]
    rownames <- rownames(all.counts)
  }
  
  #combine all counts now that the genes have been matched
  all.counts <- cbind(all.counts, counts)
}
all.counts <- as.matrix(all.counts)

#create a sce that contains all the samples and their intersecting genes
sce.S <- SingleCellExperiment(list(counts=all.counts))
rownames(sce.S) <- rownames
dim(sce.S)
```


--------------------- Linnarson 2018 data ---------------------
- this study performed scRNA-seq on nervous system tissue via 10X Chromium single cell 3' solution under homeostatic conditions

Load the Linnarson 2018 microglia loom file from http://mousebrain.org/loomfiles_level_L6.html 
```{r}
library(loomR)

lfile <- connect(filename = "/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Linnarson MG 2018/l6_r4_microglia.loom", mode = "r+")

# Extract the gene matrix
gene.names <- lfile[["row_attrs/Gene"]][]
cell_names <- lfile[["col_attrs/CellID"]][]

# extract the experimental condition (ie HC vs activated)
exp <- lfile$get.attribute.df()["Description"][,1]

full.matrix <- t(lfile$matrix[, ])
rownames(full.matrix) <- gene.names
colnames(full.matrix) <- cell_names

# convert to a sce
sce.L <- SingleCellExperiment(list(counts = full.matrix))

# label cells by experiment
sce.L$exp <- exp

# extract the HC cells
sce.L <- sce.L[,-grep("activated", exp)]
```

--------------------- Movahedi 2019 data ---------------------
- this study performed scRNA-seq on whole brain homogenates via 10X Chromium single cell 3' solution under homeostatic conditions

Load the Movahedi 2019 WT homeostatic whole brain from http://www.brainimmuneatlas.org/download.php 
```{r}
library(DropletUtils)
sce.M <- read10xCounts("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Movahedi 2019/mm10")
rownames(sce.M) <- rowData(sce.M)$Symbol

metadata <- read.csv("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Movahedi 2019/annot_K10.csv")

# reorder the metadata to match the cell order in the sce
colnames(sce.M) <- colData(sce.M)$Barcode
barcodes <- as.factor(gsub("*.-1", "", colData(sce.M)$Barcode))
metadata <- metadata[order(metadata$cell)[barcodes],]

# add the column data
sce.M$cell_type <- metadata$cluster

#metadata and sce object are of different lengths, remove the 327 cells without labels
sce.M <- sce.M[,!is.na(sce.M$cell_type)]

# extract MG
sce.M <- sce.M[,sce.M$cell_type == "Microglia"]
```

--------------------- Generate the Plot ---------------------

Load the dissociation-associated genes from van den Brink et al., 2017
```{r}
# load the function to prep the gene list
source("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/scRNAseq Workflow/Regan et al 202X/Sources/Fxn - Prep Genes.R")

# load the gene list
IEG<- read.table("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/scRNAseq Workflow/Regan et al 202X/Sources/IEGs List from van Oudenaarden (Nature Methods 2017).txt", header =F)

# convert to a character vector
IEG <- prepG(IEG, sep = "/n")
IEG <- IEG[-1]

# keep only the genes that are present in all 5 datasets
IEG <- Reduce(intersect, list(IEG, rownames(sce.L), rownames(sce.SCI), rownames(sce.Y), rownames(sce.M), rownames(sce.S)))
```

Initiate the funciton to perform QC on each dataset
```{r}
library(EnsDb.Mmusculus.v79)
library(scater)

prepSCE <- function(sce, IEGs=IEG, name) {
  #Identify mitochondrial genes 
  # Identify the chromosomal location for each gene
  rowData(sce)$CHR <- mapIds(EnsDb.Mmusculus.v79, keys=rownames(sce), 
                             column="SEQNAME", keytype="SYMBOL")
  
  # Calculate QC metrics including the percent of counts from mitochondrial genes
  sce <- calculateQCMetrics(sce, feature_controls=list(Mito=which(rowData(sce)$CHR =="MT"))) 
  
  #Remove low quality cells by UMI and gene counts thresholds and by percent of mitochondrial counts
  # identify and remove cells by low gene or UMI counts
  by.Genes <- sce$total_features_by_counts > 600
  by.UMI <- sce$total_counts > 600
  gene.UMI.keep <- by.Genes & by.UMI 
  sce <- sce[,gene.UMI.keep]
  
  # identify and remove cells by high mitochondrial gene percent
  mito <- sce$pct_counts_Mito > 10
  sce.S <- sce[,!mito]
  
  # examine the number of cells removed by each metric
  table(by.Genes, by.UMI)
  table(mito)
  
  
  #Calculate the percent of total genes from IEGs and save in a data frame
  sce <- calculateQCMetrics(sce, feature_controls = list(IEGs = which(rownames(sce) %in% IEGs)))
  
  df <- data.frame(sce$pct_counts_IEGs, rep(name, ncol(sce)), sce$pct_counts_Mito, sce$log10_total_counts)
  colnames(df) <- c("Pct", "Study", "Mito-preQC", "Logcounts-postQC")
  
  return(df)
}
```


Prepare the data frames
```{r}
df.SCI <- prepSCE(sce.SCI, name="Healthy Control")

df.M <- prepSCE(sce.M, name="Movahedi 2019")

df.L <- prepSCE(sce.L, name="Linnarsson 2018")

df.S <- prepSCE(sce.S, name="Stevens 2019")

df.Y <- prepSCE(sce.Y, name="Yang 2018")
```

--------------------- Generate the Plot ---------------------
```{r}
# combine the plots
df <- rbind(df.Y, df.L, df.S, df.M, df.SCI)

#generate the violin plot
ggplot(df, aes(x=Study, y=Pct)) + geom_violin(aes(fill = Study), color = NA) + stat_summary(fun=mean, geom="point", size=1, color="black")  + ylab("% Counts from Dissociation/FACS Genes") + xlab(NULL)+ theme(axis.text.x = element_text(angle = 45, hjust = 1),legend.position = "none") + scale_fill_manual(values=brewer.pal(5, "Dark2"))  + ylim(0,5)

#save the plot and size as needed
ggsave("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Cell Ranger 3.1.0/SCI/May 2020/BC/Hamel et al 202X/Extended Data Fig. 1d -IEGs(van den Brink list).pdf", height =7, width =4, scale = .7)
```

# Extended Data Fig. 1e UMAP by cluster 
- clusters and UMAP dimensions generated using Monocle3

Load the sce and directory
```{r}
rm(list = ls())

sce_u <- readRDS("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Cell Ranger 3.1.0/SCI/May 2020/BC/sce.RDS")
dirU <- "/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Cell Ranger 3.1.0/SCI/May 2020/BC/Hamel et al 202X/"

```

Plot the UMAP using the subset of interest
```{r}
library(scater)
library(ggplot2)

#generate the UMAP 
p <- plotUMAP(sce_u, colour_by="Clust_final_names", point_size=0.5, point_alpha=1) + geom_point(aes(fill=colData(sce_u)$Clust_final_names), colour="grey95", pch=21, stroke = .1, size =1.5, alpha = 0.6)

# convert the UMAP dots to a single image (easier handling in Illustrator)
AugmentPlot(plot=p, dpi=600)  

#save the UMAP and alter the image dimensions as needed
ggsave(filename = "Extended Data Fig. 1e -cluster.pdf", path=dirU, scale = 0.7, width=9, height=8,units = c("in", "cm", "mm"), dpi = 300, limitsize = TRUE) 
```


# S. Fig 1f Heatmaps of non-myeloid cell genes
Load and prep the sce
```{r}
rm(list=ls())
library(scater)
sce.hm <- readRDS("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Cell Ranger 3.1.0/SCI/May 2020/BC/sce.RDS")
dirH <- "/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/Cx3cr1 Data/Universal Workflow/Cell Ranger 3.1.0/SCI/May 2020/BC/Hamel et al 202X/"

# add an X to the end of the second instance of each duplicated barcode to give it a unique name (the heatmap code requires unique barcodes)
names <- make.unique(colnames(sce.hm), sep="X")

# rename the barcodes in the sce object of duplicated cell names
colnames(sce.hm) <- names
```

Prep the cell type-specific gene lists
```{r}
# load the gene lists for cell types collected from other scRNA-seq studies (see script for source articles)
source("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/scRNAseq Workflow/Gene Lists/Non-M RNA.R")

#load the function to prep the gene lists
source("/Users/reganhamel/Dropbox (Cambridge University)/Lab/Regan/SCI scRNA-seq Project/scRNAseq Workflow/Regan et al 202X/Sources/Fxn - Prep Genes.R")

# select the cell types to search for
cells <- c(Peri, Ep,Neu, Ast, Olig, MG, MC, NP, DC)
n <- c("Peri", "Ep", "Neu", "Ast", "Olig", "MG", "MC", "NP", "DC")
x <- 0
for (i in cells) {
  x <- x+1
  genes <- i
  input <- prepG(genes, sep = " ")
  assign(n[x], input)
}

# label the rows by cell type that the gene sets represent
rows <- data.frame(Genes=c(MG, MC, DC, NP,  Neu, Ast, Olig, Peri, Ep), 
                   Cell_Type=c(rep("Microglia", length(MG)),  rep("Monocyte/Monocyte-Derived", length(MC)),  rep("Dendritic Cells", length(DC)),rep("Neutrophils", length(NP)), rep("Neurons", length(Neu)), rep("Astrocytes", length(Ast)), rep("Oligodendrocytes_OPCs", length(Olig)),  rep("Pericytes", length(Peri)), rep("Ependymal_Cells", length(Ep))))

# remove duplicated genes
rows <- rows[!duplicated(rows$Genes),] 
rownames(rows) <- as.matrix(rows$Genes)

# remove genes not in the sce
rows <- rows[(rows$Genes %in% rownames(sce.hm)),]
```

Reorder and label the cell types
```{r}
library(plyr)
library(RColorBrewer)

# add empty factors to the cell_types 
sce.hm$cell_type <- factor(sce.hm$cell_type, levels=c("MG", "MCd", "DC", "NP", "Neurons", "Astrocytes", "Oligodendrocytes_OPCs", "Pericytes", "Ependymal_Cells"))

# add the full names of the cell types
sce.hm$cell_type <- revalue(sce.hm$cell_type, c("MG" ="Microglia", "MCd"="Monocyte/Monocyte-Derived", "DC"="Dendritic Cells", "NP"="Neutrophils"))
```

Generate and save the heatmap
```{r}
plotHeatmap(sce.hm, features=rows$Genes, exprs_values="logcounts", cellheight=.75, cellwidth=0.01, cluster_cols=F, cluster_rows=F, columns=order(sce.hm$cell_type), colour_columns_by="cell_type", file=paste0(dirH, "Extended Data Fig.1 CNS Cells.pdf"), fontsize=7, annotation_row=rows, show_rownames=F, center=F, symmetric=F, scale="column")

plotHeatmap(sce.hm, features=rows$Genes, exprs_values="logcounts", cellheight=.75, cellwidth=0.01, cluster_cols=F, cluster_rows=F, columns=order(sce.hm$cell_type), colour_columns_by="cell_type", file=paste0(dirH, "Extended Data Fig.1 CNS Cells.png"), fontsize=7, annotation_row=rows, show_rownames=F, center=F, symmetric=F, scale="column")

#save as png and as pdf
```
